libs/langchain/langchain/utils/math.py

```diff
@@ -46,11 +46,11 @@ def cosine_similarity_top_k(
     if len(X) == 0 or len(Y) == 0:
         return [], []
     score_array = cosine_similarity(X, Y)
-    sorted_idxs = score_array.flatten().argsort()[::-1]
-    top_k = top_k or len(sorted_idxs)
-    top_idxs = sorted_idxs[:top_k]
     score_threshold = score_threshold or -1.0
-    top_idxs = top_idxs[score_array.flatten()[top_idxs] > score_threshold]
-    ret_idxs = [(x // score_array.shape[1], x % score_array.shape[1]) for x in top_idxs]
-    scores = score_array.flatten()[top_idxs].tolist()
-    return ret_idxs, scores
+    score_array[score_array < score_threshold] = 0
+    top_k = min(top_k or len(score_array), np.count_nonzero(score_array))
```

Comment:
will np.count_nonzero be strictly less than len(score_array)? in which case latter isn't needed?

Reply:
The problem is that top_k can also be `None`, so the `or len(score_array)` is such that the other `np.count_nonzero(score_array)` always wins `min`
When `top_k` is not `None`, the `len` is just ignored.